File: tst/colors.c
#include "pico.h"
#include "../check.h"

int main (void) {
    pico_init(1);
    pico_set_window((Pico_Window){ .dim={640,480}, .fs=0, .show=1, .title="Colors Test" });
    pico_set_view_dim(NULL, &(Pico_Rel_Dim){ '!', {64, 48}, NULL });

    Pico_Anchor C = PICO_ANCHOR_C;

    puts("predefined colors");
    {
        pico_output_clear();

        pico_set_draw_color(NULL, PICO_COLOR_RED);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.1, 0.8, 0.05}, C, NULL });

        pico_set_draw_color(NULL, PICO_COLOR_GREEN);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.2, 0.8, 0.05}, C, NULL });

        pico_set_draw_color(NULL, PICO_COLOR_BLUE);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.3, 0.8, 0.05}, C, NULL });

        pico_set_draw_color(NULL, PICO_COLOR_YELLOW);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.4, 0.8, 0.05}, C, NULL });

        pico_set_draw_color(NULL, PICO_COLOR_CYAN);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.5, 0.8, 0.05}, C, NULL });

        pico_set_draw_color(NULL, PICO_COLOR_MAGENTA);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.6, 0.8, 0.05}, C, NULL });

        pico_set_draw_color(NULL, PICO_COLOR_ORANGE);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.7, 0.8, 0.05}, C, NULL });

        pico_set_draw_color(NULL, PICO_COLOR_PURPLE);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.8, 0.8, 0.05}, C, NULL });

        pico_set_draw_color(NULL, PICO_COLOR_PINK);
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.9, 0.8, 0.05}, C, NULL });

        _pico_check("colors-01");
    }

    puts("darker function");
    {
        pico_output_clear();
        for (int i=0; i<9; i++) {
            Pico_Color c = pico_color_darker(PICO_COLOR_RED, i*0.1);
            pico_set_draw_color(NULL, c);
            pico_output_draw_rect(
                &(Pico_Rel_Rect){ '%', {0.5, 0.1+i*0.1, 0.8, 0.05}, PICO_ANCHOR_C, NULL }
            );
        }
        _pico_check("colors-02");
    }

    puts("ligher function");
    {
        pico_output_clear();
        for (int i=0; i<9; i++) {
            Pico_Color c = pico_color_lighter(PICO_COLOR_BLUE, i*0.1);
            pico_set_draw_color(NULL, c);
            pico_output_draw_rect(
                &(Pico_Rel_Rect){ '%', {0.5, 0.1+i*0.1, 0.8, 0.05}, PICO_ANCHOR_C, NULL }
            );
        }
        _pico_check("colors-03");
    }

    {
        puts("negative percentages");
        pico_output_clear();

        // -dark -> light
        pico_set_draw_color(NULL, pico_color_darker(PICO_COLOR_GREEN, -0.50));
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.25, 0.5, 0.15}, PICO_ANCHOR_C, NULL }
        );

        // green
        pico_set_draw_color(NULL, PICO_COLOR_GREEN);
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.50, 0.5, 0.15}, PICO_ANCHOR_C, NULL }
        );

        // -light -> dark
        pico_set_draw_color(NULL, pico_color_lighter(PICO_COLOR_GREEN, -0.50));
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.75, 0.5, 0.15}, PICO_ANCHOR_C, NULL }
        );

        _pico_check("colors-04");
    }

    puts("mix function");
    {
        pico_output_clear();

        // red + blue = purple
        Pico_Color c1 = pico_color_mix(PICO_COLOR_RED, PICO_COLOR_BLUE);
        pico_set_draw_color(NULL, c1);
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.15, 0.8, 0.1}, PICO_ANCHOR_C, NULL }
        );

        // red + green = olive
        Pico_Color c2 = pico_color_mix(PICO_COLOR_RED, PICO_COLOR_GREEN);
        pico_set_draw_color(NULL, c2);
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.35, 0.8, 0.1}, PICO_ANCHOR_C, NULL }
        );

        // blue + green = teal
        Pico_Color c3 = pico_color_mix(PICO_COLOR_BLUE, PICO_COLOR_GREEN);
        pico_set_draw_color(NULL, c3);
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.55, 0.8, 0.1}, PICO_ANCHOR_C, NULL }
        );

        // black + white = gray
        Pico_Color c4 = pico_color_mix(PICO_COLOR_BLACK, PICO_COLOR_WHITE);
        pico_set_draw_color(NULL, c4);
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.75, 0.8, 0.1}, PICO_ANCHOR_C, NULL }
        );

        _pico_check("colors-05");
    }

    // mirrors Lua test for '%' and '!' color table formats
    {
        puts("percentage color format");
        pico_output_clear();

        // {'%', r=1, g=0, b=0} -> (255, 0, 0)
        pico_set_draw_color(NULL, (Pico_Color){255, 0, 0, 0xFF});
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.15, 0.8, 0.1}, C, NULL }
        );

        // {'%', r=0, g=1, b=0} -> (0, 255, 0)
        pico_set_draw_color(NULL, (Pico_Color){0, 255, 0, 0xFF});
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.30, 0.8, 0.1}, C, NULL }
        );

        // {'%', r=0, g=0, b=1} -> (0, 0, 255)
        pico_set_draw_color(NULL, (Pico_Color){0, 0, 255, 0xFF});
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.45, 0.8, 0.1}, C, NULL }
        );

        // {'%', r=0.5, g=0.5, b=0.5} -> (127, 127, 127)
        pico_set_draw_color(NULL, (Pico_Color){127, 127, 127, 0xFF});
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.60, 0.8, 0.1}, C, NULL }
        );

        // {'!', r=255, g=255, b=0} -> (255, 255, 0)
        pico_set_draw_color(NULL, (Pico_Color){255, 255, 0, 0xFF});
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.75, 0.8, 0.1}, C, NULL }
        );

        // {r=0, g=255, b=255} -> (0, 255, 255)
        pico_set_draw_color(NULL, (Pico_Color){0, 255, 255, 0xFF});
        pico_output_draw_rect(
            &(Pico_Rel_Rect){ '%', {0.5, 0.90, 0.8, 0.1}, C, NULL }
        );

        _pico_check("colors-06");
    }

    puts("hex color format");
    {
        pico_output_clear();

        pico_set_draw_color(NULL, pico_color_hex(0xFF0000));
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.15, 0.8, 0.1}, C, NULL });

        pico_set_draw_color(NULL, pico_color_hex(0x00FF00));
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.30, 0.8, 0.1}, C, NULL });

        pico_set_draw_color(NULL, pico_color_hex(0x0000FF));
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.45, 0.8, 0.1}, C, NULL });

        pico_set_draw_color(NULL, pico_color_hex(0x7F7F7F));
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.60, 0.8, 0.1}, C, NULL });

        pico_set_draw_color(NULL, pico_color_hex(0xFFFF00));
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.75, 0.8, 0.1}, C, NULL });

        pico_set_draw_color(NULL, pico_color_hex(0x00FFFF));
        pico_output_draw_rect(&(Pico_Rel_Rect){ '%', {0.5, 0.90, 0.8, 0.1}, C, NULL });

        _pico_check("colors-06");
    }

    puts("hex color conversion");
    {
        Pico_Color c;

        c = pico_color_hex(0xFF0000);
        assert(c.r==255 && c.g==0 && c.b==0);

        c = pico_color_hex(0x00FF00);
        assert(c.r==0 && c.g==255 && c.b==0);

        c = pico_color_hex(0x0000FF);
        assert(c.r==0 && c.g==0 && c.b==255);

        c = pico_color_hex(0x336699);
        assert(c.r==0x33 && c.g==0x66 && c.b==0x99);

        c = pico_color_hex(0x000000);
        assert(c.r==0 && c.g==0 && c.b==0);

        c = pico_color_hex(0xFFFFFF);
        assert(c.r==255 && c.g==255 && c.b==255);

        c = pico_color_hex(0xFF000080);
        assert(c.r==255 && c.g==0 && c.b==0 && c.a==0x80);

        c = pico_color_hex(0x33669900);
        assert(c.r==0x33 && c.g==0x66 && c.b==0x99 && c.a==0);

        c = pico_color_hex(0xFFFFFFFF);
        assert(c.r==255 && c.g==255 && c.b==255 && c.a==255);
    }

    puts("color alpha function");
    {
        pico_set_show_color(NULL, PICO_COLOR_WHITE);
        pico_output_clear();
        Pico_Color buffer[] = {
            pico_color_alpha(PICO_COLOR_RED, 0xFF),
            pico_color_alpha(PICO_COLOR_RED, 0x80),
            PICO_COLOR_TRANSPARENT,
        };
        Pico_Rel_Rect r = {
            '%', {0.5, 0.5, 0, 0}, PICO_ANCHOR_C, NULL
        };
        pico_output_draw_buffer(
            "alpha", (Pico_Abs_Dim){3,1}, buffer, &r
        );
        _pico_check("colors-07");
    }

    pico_init(0);
    return 0;
}

File: lua/tst/colors.lua
require 'pico.check'

pico.init(true)
pico.set.window { title="Colors Test" }
pico.set.layer("window")
pico.set.scene { dim={'!', w=640, h=480} }
pico.set.layer("world")
pico.set.scene { dim={'!', w=64, h=48} }

do
    print("predefined colors")
    pico.output.clear()

    pico.set.pencil { color='red' }
    pico.output.draw.rect({'%', x=0.5, y=0.1, w=0.8, h=0.05})

    pico.set.pencil { color='green' }
    pico.output.draw.rect({'%', x=0.5, y=0.2, w=0.8, h=0.05})

    pico.set.pencil { color='blue' }
    pico.output.draw.rect({'%', x=0.5, y=0.3, w=0.8, h=0.05})

    pico.set.pencil { color='yellow' }
    pico.output.draw.rect({'%', x=0.5, y=0.4, w=0.8, h=0.05})

    pico.set.pencil { color='cyan' }
    pico.output.draw.rect({'%', x=0.5, y=0.5, w=0.8, h=0.05})

    pico.set.pencil { color='magenta' }
    pico.output.draw.rect({'%', x=0.5, y=0.6, w=0.8, h=0.05})

    pico.set.pencil { color='orange' }
    pico.output.draw.rect({'%', x=0.5, y=0.7, w=0.8, h=0.05})

    pico.set.pencil { color='purple' }
    pico.output.draw.rect({'%', x=0.5, y=0.8, w=0.8, h=0.05})

    pico.set.pencil { color='pink' }
    pico.output.draw.rect({'%', x=0.5, y=0.9, w=0.8, h=0.05})

    pico.check("colors-01")
end

do
    print("darker function")
    pico.output.clear()
    for i = 0, 8 do
        local c = pico.color.darker('red', i*0.1)
        pico.set.pencil { color=c }
        pico.output.draw.rect({'%', x=0.5, y=0.1+i*0.1, w=0.8, h=0.05})
    end
    pico.check("colors-02")
end

do
    print("lighter function")
    pico.output.clear()
    for i = 0, 8 do
        local c = pico.color.lighter('blue', i*0.1)
        pico.set.pencil { color=c }
        pico.output.draw.rect({'%', x=0.5, y=0.1+i*0.1, w=0.8, h=0.05})
    end
    pico.check("colors-03")
end

do
    print("negative percentages")
    pico.output.clear()

    -- -dark -> light
    pico.set.pencil { color=pico.color.darker('green', -0.50) }
    pico.output.draw.rect({'%', x=0.5, y=0.25, w=0.5, h=0.15})

    -- green
    pico.set.pencil { color='green' }
    pico.output.draw.rect({'%', x=0.5, y=0.50, w=0.5, h=0.15})

    -- -light -> dark
    pico.set.pencil { color=pico.color.lighter('green', -0.50) }
    pico.output.draw.rect({'%', x=0.5, y=0.75, w=0.5, h=0.15})

    pico.check("colors-04")
end

do
    print("mix function")
    pico.output.clear()

    -- red + blue = purple
    local c1 = pico.color.mix('red', 'blue')
    pico.set.pencil { color=c1 }
    pico.output.draw.rect({'%', x=0.5, y=0.15, w=0.8, h=0.1})

    -- red + green = olive
    local c2 = pico.color.mix('red', 'green')
    pico.set.pencil { color=c2 }
    pico.output.draw.rect({'%', x=0.5, y=0.35, w=0.8, h=0.1})

    -- blue + green = teal
    local c3 = pico.color.mix('blue', 'green')
    pico.set.pencil { color=c3 }
    pico.output.draw.rect({'%', x=0.5, y=0.55, w=0.8, h=0.1})

    -- black + white = gray
    local c4 = pico.color.mix('black', 'white')
    pico.set.pencil { color=c4 }
    pico.output.draw.rect({'%', x=0.5, y=0.75, w=0.8, h=0.1})

    pico.check("colors-05")
end

do
    print("percentage color format")
    pico.output.clear()

    -- {'%', r, g, b} format - percentage (0.0-1.0)
    pico.set.pencil { color={'%', r=1, g=0, b=0} }
    pico.output.draw.rect({'%', x=0.5, y=0.15, w=0.8, h=0.1})

    pico.set.pencil { color={'%', r=0, g=1, b=0} }
    pico.output.draw.rect({'%', x=0.5, y=0.30, w=0.8, h=0.1})

    pico.set.pencil { color={'%', r=0, g=0, b=1} }
    pico.output.draw.rect({'%', x=0.5, y=0.45, w=0.8, h=0.1})

    -- 50% gray
    pico.set.pencil { color={'%', r=0.5, g=0.5, b=0.5} }
    pico.output.draw.rect({'%', x=0.5, y=0.60, w=0.8, h=0.1})

    -- {'!', r, g, b} format - explicit absolute (0-255)
    pico.set.pencil { color={'!', r=255, g=255, b=0} }
    pico.output.draw.rect({'%', x=0.5, y=0.75, w=0.8, h=0.1})

    -- {r, g, b} format - implicit absolute (backward compatible)
    pico.set.pencil { color={r=0, g=255, b=255} }
    pico.output.draw.rect({'%', x=0.5, y=0.90, w=0.8, h=0.1})

    pico.check("colors-06")
end

do
    print("hex color format")
    pico.output.clear()

    pico.set.pencil { color=0xFF0000 }
    pico.output.draw.rect({'%', x=0.5, y=0.15, w=0.8, h=0.1})

    pico.set.pencil { color=0x00FF00 }
    pico.output.draw.rect({'%', x=0.5, y=0.30, w=0.8, h=0.1})

    pico.set.pencil { color=0x0000FF }
    pico.output.draw.rect({'%', x=0.5, y=0.45, w=0.8, h=0.1})

    pico.set.pencil { color=0x7F7F7F }
    pico.output.draw.rect({'%', x=0.5, y=0.60, w=0.8, h=0.1})

    pico.set.pencil { color=0xFFFF00 }
    pico.output.draw.rect({'%', x=0.5, y=0.75, w=0.8, h=0.1})

    pico.set.pencil { color=0x00FFFF }
    pico.output.draw.rect({'%', x=0.5, y=0.90, w=0.8, h=0.1})

    pico.check("colors-06")
end

do
    print("hex color conversion")

    local c = pico.color.darker(0xFF0000, 0)
    assert(c.r==255 and c.g==0 and c.b==0)

    c = pico.color.darker(0x00FF00, 0)
    assert(c.r==0 and c.g==255 and c.b==0)

    c = pico.color.darker(0x0000FF, 0)
    assert(c.r==0 and c.g==0 and c.b==255)

    c = pico.color.darker(0x336699, 0)
    assert(c.r==0x33 and c.g==0x66 and c.b==0x99)

    c = pico.color.darker(0x000000, 0)
    assert(c.r==0 and c.g==0 and c.b==0)

    c = pico.color.darker(0xFFFFFF, 0)
    assert(c.r==255 and c.g==255 and c.b==255)
end

do
    print("color alpha function")
    pico.set.effect { color='white' }
    pico.output.clear()
    local pixmap = {
        {
            pico.color.alpha('red', 0xFF),
            pico.color.alpha('red', 0x80),
            'transparent',
        },
    }
    local r = {'%', x=0.5, y=0.5, w=0, h=0}
    pico.output.draw.pixmap("alpha", pixmap, r)
    pico.check("colors-07")
end

do
    print("pixmap with mixed color formats")
    pico.set.effect { color='white' }
    pico.output.clear()
    local pixmap = {
        { 'red',                 0x00FF00, { r=0,   g=0,   b=255 } },
        { {'%', r=1, g=1, b=0},  'cyan',   {'!', r=255, g=0, b=255} },
        { 0x000000,              'black',  pico.color.alpha('red', 0x80) },
    }
    local r = {'%', x=0.5, y=0.5, w=0, h=0}
    pico.output.draw.pixmap("mixed", pixmap, r)
    pico.check("colors-0X")
end

-- window letterbox color: shrink dst to expose letterbox area
do
    do
        pico.set.effect { color='black' }
        pico.set.scene { target={'%', x=1, y=1, w=0.5, h=0.5, anchor='SE'} }
    end
    do
        print("window color default: gray {0x77,0x77,0x77,0xFF}")
        pico.set.layer("window")
        local c = pico.get.effect().color
        pico.set.layer("world")
        assert(c.r==0x77 and c.g==0x77 and c.b==0x77 and c.a==0xFF)
        pico.output.clear()
        pico.set.pencil { color='white' }
        pico.output.draw.rect({'%', x=0.5, y=0.5, w=0.5, h=0.5})
        pico.check("colors-08")
    end
    do
        print("window color: red letterbox")
        pico.set.layer("window")
        pico.set.effect { color={'!', r=0xFF, g=0, b=0, a=0xFF} }
        local c = pico.get.effect().color
        pico.set.layer("world")
        assert(c.r==0xFF and c.g==0 and c.b==0 and c.a==0xFF)
        pico.output.clear()
        pico.output.draw.rect({'%', x=0.5, y=0.5, w=0.5, h=0.5})
        pico.check("colors-09")
    end
    do
        print("window color: green via bulk")
        pico.set.layer("window")
        pico.set.effect { color={'!', r=0, g=0xFF, b=0, a=0xFF} }
        local c = pico.get.effect().color
        pico.set.layer("world")
        assert(c.r==0 and c.g==0xFF and c.b==0 and c.a==0xFF)
        pico.output.clear()
        pico.output.draw.rect({'%', x=0.5, y=0.5, w=0.5, h=0.5})
        pico.check("colors-10")
    end
    do
        print("window color: alpha preserved by storage")
        pico.set.layer("window")
        pico.set.effect { color={'!', r=0, g=0, b=0xFF, a=0x80} }
        local c = pico.get.effect().color
        pico.set.layer("world")
        assert(c.a == 0x80)
    end
end

pico.init(false)
